README
Build instructions can be found in INSTALL

- Supported Games

  This table displays which of the main actions are
supported for each game.

Action   1 2 3 4 5 6 7 75 8 9 95 10 103 105 11 12 123 125 128 13 14 143 15 16 165 17 18 185 19
thanm x  - - - - - - y  - y y  y  y   y  -  y  y   -   y   y  y  y   y  y  y   y  y  y  y   y
thanm l  - - - - - y y  - y y  y  y   y  -  y  y   -   y   y  y  y   y  y  y   y  y  y  y   y
thdat x  y y y y y y y  y y y  y  y   y  y  y  y   y   y   y  y  y   y  y  y   y  y  y  y   y
thdat c  y y y y y y y  y y y  y  y   y  y  y  y   y   y   y  y  y   y  y  y   y  y  y  y   y
thecl d  - - - - - y y  - y y  y  y   y  -  y  y   -   y   y  y  y   y  y  y   y  y  y  y   y
thecl c  - - - - - y y  - y y  y  y   y  -  y  y   -   y   y  y  y   y  y  y   y  y  y  y   y
thmsg d  - - - - - y y  - y y  y  y   -  -  y  y   -   y   y  y  y   y  y  y   y  y  y  y   y
thmsg c  - - - - - y y  - y y  y  y   -  -  y  y   -   y   y  y  y   y  y  y   y  y  y  y   y
thstd d  - - - - - y y  - y y  y  y   y  -  y  y   -   y   y  y  y   y  y  y   y  y  y  y   y
thstd c  - - - - - y y  - y y  y  y   y  -  y  y   -   y   y  y  y   y  y  y   y  y  y  y   y


NEWS
What's new in thtk master
=====================

#### thtk
- Improvements to the CMake build process.
- thlzss.h and thcrypt.h are now part of the API.
- We've set up GitHub Actions for automatic builds.

#### thanm
- New thanm spec format. See <https://github.com/thpatch/thtk/pull/86> for more
  details. The old format is not supported anymore. Use thanm.old for it.
- Game version now must be specified, just as with other tools.
- Support for TH18, TH185, TH19 (full) has been added.
- Support th143/bestshot.anm
- Two new options (-u and -uu) were added to help with rebuilding TH19 files
  bit-perfectly. See documentation for more details.
- A new option (-X) to extract images from multiple ANM files at once.

#### thanm.old
- Will be removed in the next release.
- Support for TH18, TH185, TH19 (trial) has been added.
  TH19 creation won't work, though. Use the new thanm instead.
- You can now specify the version for the -l command.
  This is required for correct dumping of TH18, TH185.
- Support th143/bestshot.anm

#### thecl
- Fix and/or add the following instruction formats:
    th06 71-74, 90
    th08 14, 19, 72, 84-86, 118, 119, 152
    th13 18. 19, 20, 24
- The eclmap format has been changed. It's documented in the thecl manual.
  A converter is available in contrib/eclmapcvt.awk.
- Allow multiple mnemonics for the same instruction.
- Allow eclmap signatures for timeline instructions.
- The parser grammar has been reworked to be completely unambiguous.
- Control statements now require parenthesis around expressions.
  E.g. if (x) goto y; instead of if x goto y;
- More escape sequences are supported: \r \n \0.
- Unary plus operator added.
- Expressions are compile-time evaluated, and constant expressions are allowed
  in more places.
- Added #line directive.
- Added insdef statement for defining new instructions.
- Support for TH18, TH185, TH19 has been added.
- Numerous bug-fixes.
- Add -x option for outputting address information for instructions.
- Add -C option for changing the current directory after opening the archive.
- Add -j option for converting strings between Shift-JIS and UTF-8.

#### thdat
- Automatic version detection now supports th165, th17, th18, th185, th19.
- Detection works better with renamed files.
- Support for TH18, TH185, TH19 has been added.
- Improve extraction speed.
- Add -g option to enable globing on extracted filenames.
  Example: thdat -gx18 th18.dat "*.ecl"
- Support for older Tasogare Frontier games added:
  IaMP, Super Marisa Land, MegaMari, Higurashi Daybreak, PatchCon

#### thmsg
- Support for TH18, TH185, TH19 has been added.

#### thstd
- Support for TH18, TH185, TH19 has been added.

Please submit an issue at GitHub (https://github.com/thpatch/thtk/issues) if
you find a bug.


-------------------------------------------------------------------------------

What's new in thtk 12
=====================

A lot of improvements for thecl. If you want to recompile ZUN's files
bit-for-bit, make sure you use the simple creation flag (-s) introduced in this
version.

#### thanm
- Support for TH17 has been added.

#### thecl
- Scan and parse negative (-val) function. (th13+ only)
- Complete support for th16.5.
- Look up ≥th10 instruction formats even if they have no parameters.
- Print the total parameter size for unknown ≥th10 instructions.
- th10: Add neqf to format table.
- Print error and quit if expressions contain invalid types.
- Allow floats in the form 1.f
- Allow floats in the form +1f (and +1.2f and +1.f)
- Allow integers in the form +123
- Add sin, cos and sqrt expressions.
- Allow sub calls by name with parameters cast by type.
- Add proper if-else blocks.
- Add while and (simple) switch blocks.
- Add global definitions.
- Add missing and fix existing EoSD instructions.
- th06: Check if instruction's opcode is higher than the max usable by the game
- Add bitwise operators.
- Fix argument types for a handful of th12 and th13 instructions.
- Support for TH17 has been added.
- Add escaping quotes and backslashes in strings.
- Add ZUN-style difficulty switches for writing custom ECL
- Don't force typecasts for instruction calls
- Add syntax for writing degree-floats (auto converts value to radians).
- Add syntax for writing binary integer literals.
- Add forced sub calls by name.
- Allow var declarations anywhere in the sub.
- Add variable initialization.
- Automatically add missing ins_10 at the end of a sub.
- Better difficulty switch expression.
- Add times loop.
- Add simple creation mode.
- Added typed variables.
- Improved code blocks.
- Add shorthand operators. (e.g. +=)
- Constant math optimization pass.
- Add #include directive
- Add #eclmap directive
- Add single-line comments.
- Add proper pre-th10 timeline dumping.
- Add #ins directive
- Typed functions, return values.
- Variable scoping.
- Parse true/false as integer 1/0 respectively.
- Add relative time labels (i.e. +60:). These are the default format when
  dumping. The total absolute value is appended as a comment.
- Add inline sub support. Prepend the keyword "inline" to the sub type to make
  every call to that sub inline. Inline subs cannot call themselves.
- Gotos will automatically assign the same time as to where the label is
  located if no time is specified.

#### thdat
- Support for TH17 has been added.
- Add support for th09e.dat.

#### thmsg
- Support for TH17 has been added.

#### thstd
- Support for TH17 has been added.

Please submit an issue at GitHub (https://github.com/thpatch/thtk/issues) if
you find a bug.


-------------------------------------------------------------------------------

What's new in thtk 11
=====================

This release adds thstd, a new tool to dump and recreate the background
scripts found in the *.std files. Further changes to existing tools:

#### thanm
In preparation of thcrap's upcoming support for ANM header patching, more
thorough reverse-engineering went into the various ANM structures.

expr.c
/*
 * Redistribution and use in source and binary forms, with
 * or without modification, are permitted provided that the
 * following conditions are met:
 *
 * 1. Redistributions of source code must retain this list
 *    of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce this
 *    list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the
 *    distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 */
#include <config.h>
#include <stddef.h>
#include "thecl.h"
#include "ecsparse.h"
#include "expr.h"

typedef struct {
    int symbol;
    int symbols[2];
} alternative_t;

static const alternative_t
th10_alternatives[] = {
    { ADD,      { ADDI, ADDF } },
    { SUBTRACT, { SUBTRACTI, SUBTRACTF } },
    { MULTIPLY, { MULTIPLYI, MULTIPLYF } },
    { DIVIDE,   { DIVIDEI, DIVIDEF } },
    { EQUAL,    { EQUALI, EQUALF } },
    { INEQUAL,  { INEQUALI, INEQUALF } },
    { LT,       { LTI, LTF } },
    { LTEQ,     { LTEQI, LTEQF } },
    { GT,       { GTI, GTF } },
    { GTEQ,     { GTEQI, GTEQF } },
    { 0,        { 0, 0 } }
};

static const expr_t
th10_expressions[] = {
    /* The program checks against the number of params, as well as the
     * requested stack depth, and does the replacements. */
    /* p0 is the first param, p1 the second ... */
    /* s0 is the previous instruction, s1 the one previous to s0 ... */

    /*SYM         ID  RET     P  A    S   DISP */
    { GOTO,       12,   0, "oS", 0,  "S",           "goto p0 @ p1" },
    { UNLESS,     13,   0, "oS", 1,  "S", "unless s0 goto p0 @ p1" },
    { IF,         14,   0, "oS", 1,  "S",     "if s0 goto p0 @ p1" },

    { LOADI,      42, 'S',  "S", 0, NULL, "p0" },
    { ASSIGNI,    43,   0,  "S", 1,  "S", "p0 = s0" },
    { LOADF,      44, 'f',  "f", 0, NULL, "p0" },
    { ASSIGNF,    45,   0,  "f", 1,  "f", "p0 = s0" },

    { ADDI,       50, 'S', NULL, 2, "SS", "s1 + s0" },
    { ADDF,       51, 'f', NULL, 2, "ff", "s1 + s0" },
    { SUBTRACTI,  52, 'S', NULL, 2, "SS", "s1 - s0" },
    { SUBTRACTF,  53, 'f', NULL, 2, "ff", "s1 - s0" },
    { MULTIPLYI,  54, 'S', NULL, 2, "SS", "s1 * s0" },
    { MULTIPLYF,  55, 'f', NULL, 2, "ff", "s1 * s0" },
    { DIVIDEI,    56, 'S', NULL, 2, "SS", "s1 / s0" },
    { DIVIDEF,    57, 'f', NULL, 2, "ff", "s1 / s0" },
    { MODULO,     58, 'S', NULL, 2, "SS", "s1 % s0" },
    { EQUALI,     59, 'S', NULL, 2, "SS", "s1 == s0" },
    { EQUALF,     60, 'S', NULL, 2, "ff", "s1 == s0" },
    { INEQUALI,   61, 'S', NULL, 2, "SS", "s1 != s0" },
    { INEQUALF,   62, 'S', NULL, 2, "ff", "s1 != s0" },
    { LTI,        63, 'S', NULL, 2, "SS", "s1 < s0" },
    { LTF,        64, 'S', NULL, 2, "ff", "s1 < s0" },
    { LTEQI,      65, 'S', NULL, 2, "SS", "s1 <= s0" },
    { LTEQF,      66, 'S', NULL, 2, "ff", "s1 <= s0" },
    { GTI,        67, 'S', NULL, 2, "SS", "s1 > s0" },
    { GTF,        68, 'S', NULL, 2, "ff", "s1 > s0" },
    { GTEQI,      69, 'S', NULL, 2, "SS", "s1 >= s0" },
    { GTEQF,      70, 'S', NULL, 2, "ff", "s1 >= s0" },
    { NOT,        71, 'S', NULL, 1, "SS", "!s0" },
/*  { XXX,        72,   0, NULL, 0, NULL, NULL },*/
    { OR,         73, 'S', NULL, 2, "SS", "s1 || s0" },
    { AND,        74, 'S', NULL, 2, "SS", "s1 && s0" },
    { XOR,        75, 'S', NULL, 2, "SS", "s1 ^ s0" },
    { 0,           0,   0, NULL, 0, NULL, NULL }
};

static const expr_t*
expr_get_table(unsigned int version)
{
    if (   version != 10
        && version != 11
        && version != 12
        && version != 125
        && version != 128
        && version != 13
        && version != 14
        && version != 143)
        return NULL;
    return th10_expressions;
}

const expr_t*
expr_get_by_symbol(
    unsigned int version,
    int symbol)
{
    const expr_t* table = expr_get_table(version);

    while (table->symbol) {
        if (table->symbol == symbol)
            return table;
        ++table;
    }

    return NULL;
}

const expr_t*
expr_get_by_id(
    unsigned int version,
    int id)
{
    const expr_t* table = expr_get_table(version);

    while (table->symbol) {
        if (table->id == id)
            return table;
        ++table;
    }

    return NULL;
}

int
expr_is_leaf(
    unsigned int version,
    int id)
{
    const expr_t* expr = expr_get_by_id(version, id);

    if (!expr)
        return 0;

    return expr->stack_arity == 0 &&
           expr->return_type != 0;
}
